blockauth/serializers/wallet_serializers.py

```
@@ -138,6 +138,68 @@ def authenticate_user(self):
+        signature = data.get("signature")
+        request = self.context.get("request")
+
+        # 1. Verify signature — replay protection, nonce, timestamp all handled here
+        try:
+            authenticator = WalletAuthenticator()
+            if not authenticator.verify_signature(wallet_address, message, signature):
+                raise ValidationError(
+                    detail={"signature": "Invalid signature. Signature verification failed."}, code="INVALID_SIGNATURE"
+                )
+        except ValueError as e:
+            raise ValidationError(detail={"message": str(e)}, code="INVALID_SIGNATURE")
```

Ruff (0.15.9)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)